src/libraries/Microsoft.Extensions.Diagnostics/src/Metrics/ListenerSubscription.cs

```diff
@@ -0,0 +1,324 @@
+    {
+        private readonly MeterListener _meterListener;
+        private readonly IMetricsListener _metricsListener;
+#if NET
+        private readonly ConditionalWeakTable<Instrument, InstrumentStatus> _instruments = new();
+#else
+        // TODO: Can't enumerate ConditionalWeakTable in .NET Standard 2.0. How do we clean up the instruments?
+        private readonly Dictionary<Instrument, InstrumentStatus> _instruments = new();
+#endif
+        private IList<InstrumentRule> _rules = Array.Empty<InstrumentRule>();
+        private bool _disposed;
+
+        internal ListenerSubscription(IMetricsListener metricsListener)
+        {
+            _metricsListener = metricsListener;
+
+            _meterListener = new MeterListener();
+        }
+
+        public void Initialize()
+        {
+            _meterListener.InstrumentPublished = InstrumentPublished;
+            _meterListener.MeasurementsCompleted = MeasurementsCompleted;
+            var handlers = _metricsListener.GetMeasurementHandlers();
+            _meterListener.SetMeasurementEventCallback(handlers.ByteHandler);
+            _meterListener.SetMeasurementEventCallback(handlers.ShortHandler);
+            _meterListener.SetMeasurementEventCallback(handlers.IntHandler);
+            _meterListener.SetMeasurementEventCallback(handlers.LongHandler);
+            _meterListener.SetMeasurementEventCallback(handlers.FloatHandler);
+            _meterListener.SetMeasurementEventCallback(handlers.DoubleHandler);
+            _meterListener.SetMeasurementEventCallback(handlers.DecimalHandler);
+            _meterListener.Start();
+            _metricsListener.Initialize(this);
+        }
+
+        private void InstrumentPublished(Instrument instrument, MeterListener _)
+        {
+            lock (_instruments)
+            {
+                if (_disposed)
+                {
+                    return;
+                }
+
+                if (_instruments.TryGetValue(instrument, out var _))
+                {
+                    Debug.Assert(false, "InstrumentPublished called twice for the same instrument");
+                    return;
+                }
+
+                var status = new InstrumentStatus();
+                _instruments.Add(instrument, status);
+                RefreshInstrument(instrument, status);
+            }
+        }
+
+        // Called when we call DisableMeasurementEvents, like when a rule is disabled.
+        private void MeasurementsCompleted(Instrument instrument, object? state)
+        {
+            lock (_instruments)
+            {
+                if (_disposed)
+                {
+                    return;
+                }
+
+                if (!_instruments.TryGetValue(instrument, out var status))
+                {
+                    Debug.Assert(false, "MeasurementsCompleted called for an instrument that was never published");
+                    return;
+                }
+                status.Enabled = false;
```

Comment:
>status.Enabled = false; [](http://example.com/codeflow?start=16&length=23)

as @noahfalk suggested we can track only the active instruments and remove them when they are completed.